### PR TITLE
[Lang] Add 2x2/3x3 matrix solve with Guass elimination

### PR DESCRIPTION
**Pull request by FantasyVR**
Related issue = #2293 #2208

### CODE REVIEW

**Review by Hanke98 — COMMENTED** [review]
Good work! 
Sorry for the delayed reply. I would like to suggest adding tests about `_gauss_elimination` and `_forward_elimination` respectively.

File: misc/solve_3x3.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+ti.init()
+
+
+@ti.kernel
+def test_solve_3x3():
+    A = ti.Matrix([[3.0, 2.0, -4.0], [2.0, 3.0, 3.0], [5.0, -3, 1.0]])
+    b = ti.Vector([3.0, 15.0, 14.0])
+    x = A.solve(b)
+    print(x)
+
```
[the remark this answers is not in the document]
Comment (FantasyVR) [reply misc/solve_3x3.py]: Thanks for your review. I rm this test and add more tests under the test folder.

File: python/taichi/_funcs.py [hunk 2]
```diff
@@ -515,4 +515,79 @@ def sym_eig(A, dt=None):
...
+                    Ab[i, col], Ab[0, col] = Ab[0, col], Ab[i, col]
+            elif max_row == 1:
+                for col in static(range(ncol)):
+                    Ab[i, col], Ab[1, col] = Ab[1, col], Ab[i, col]
+            else:
+                for col in static(range(ncol)):
+                    Ab[i, col], Ab[2, col] = Ab[2, col], Ab[i, col]
+        for j in static(range(i + 1, nrow)):
+            scale = Ab[j, i] / Ab[i, i]
+            Ab[j, i] = 0.0
```
[the remark this answers is not in the document]
Comment (FantasyVR) [reply python/taichi/_funcs.py:542]: Good point.

**Comment by FantasyVR** [comment]
>  I would like to suggest adding tests about `_gauss_elimination` and `_forward_elimination` respectively.

Hi @Hanke98, I add tests for 2x2 and 3x3 matrix solve and the `gauss elimination` and `back substitution` are implemented in one place. And I split the implementation of 2x2 and 3x3 matrix solve because of the limitation of the dynamic index problem.  The `divide by zero` problems is prevented using assertions.

**Review by neozhaoliang — APPROVED** [review]
Hi, thanks for this contribution, learned a new feature of pytest~

**Comment by neozhaoliang** [comment]
Would you please check the results for 

```
    @ti.kernel
    def solve_2x2():
        A[None] = ti.Matrix([[a00, 1.0], [1.0, 1.001]])
        b[None] = ti.Vector([3.0, 15.0])
        x[None] = ti.solve(A[None], b[None])
```
and

```
@pytest.mark.parametrize('a00', [i/10 for i in range(10, 13)])
@test_utils.test(default_fp=ti.f32, fast_math=False)
def test_solve_2x2_f32(a00):
    _test_solve_2x2(ti.f32, a00)
```

**Comment by FantasyVR** [comment]
> These equations are solvable in numpy, but reports the singular matrix error in the test.

Thanks for your review. I rerun the test case your provided, but I couldn't reproduce the failed test cases. Could you show me the value of `a00` when  the test case fail?

**Pull request merged**